src/FreeAge/common/type_stats.hpp

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 The FreeAge authors
+// This file is part of FreeAge, licensed under the new BSD license.
+// See the COPYING file in the project root for the license text.
+
+#pragma once
+
+#include <QRect>
+#include <QSize>
+
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/resources.hpp"
+
+static constexpr u32 typeStatsVersion = 1;
+
+enum class AttackType {
+  None,
+  /// TODO: rename ?
+  Default,
+  /// Full damage to all affected in the area of effect.
+  HomogeneousAreaOfEffect,
+  /// Full damage to primary target and 0.5 to all others in the area of effect.
+  TrampleHalf, // War Elephants, Battle Elephant
+  /// Full damage to primary target and 5 unaffected by armor damage in the area of effect.
+  TrampleFivePure, // Cataphracts, Druzhina
+  /// First projectile does full damage and other projectiles deal 3 pierce damage and 0 melee damage.
+  MultipleProjectiles, // Chu Ko Nu, Kipchak
+  
+  TownCenter, Castle, Tower, BombardTower
+};
+
+enum class GarrisonType {
+  None,
+  Units, // TransportShip ?
+  VillagersAndFootArchersAndInfantry, // TownCenter and towers
```

Comment:
More compact `VillagersAndFootSoldiers`? It applies also for Rams I think. 
There's also the Siege Tower, which, according to [aoewiki](https://ageofempires.fandom.com/wiki/Garrison) can only garrison FootSoldiers but no vills.

Reply:
There's more nasty details, e.g. monks and petards cannot garrison in siege rams/towers.

Reply:
I wiil add `VillagersAndFootSoldiers` and `FootSoldiers`, but I don't think the monks and petards will cause any problem because they are not considered soldiers

Reply:
But at least monks can garrison in TCs and towers can't they? 
But probably these details are also irrelevant at this stage of the project, so feel free to ignore.

Reply:
Yes, you are right! No, I don't think they are irrelevant, they help having to revisit and refactor code in later stages (which could be painful sometimes).

Then I think they could be considered "Villagers" in the context of garrison (I'm not sure for that either, does it work?)

Reply:
> Then I think they could be considered "Villagers" in the context of garrison (I'm not sure for that either, does it work?)

Not really, as vills can garrison in rams, while monks can't. 

Just found this page, it seems like a great resource, not only for garrisoning: 
https://www.aoe2sheet.info/knowledge_base.html#garrisoning

OT: 
Holy shit, have a look at the monk mechanics!

Reply:
@grst How about now?
```c++
enum class GarrisonType {
  /// No unit can be garrison.
  None,
  /// All units
  AllUnits, // Transport ships
  /// Villagers, monks and foot soldiers only.
  VillagersAndMonksAndFootSoldiers, // TownCenter and towers
  /// Villagers and foot soldiers only.
  VillagersFootSoldiers, // Rams and Siege Towers
  /// All units except siege units.
  NonSiege, // Castle
  /// Villagers only.
  Villagers, // Khmer houses
  /// Units from production buildings can be trained directly into garrison,
  /// but not garrison from outside.
  Production,
  /// Same as Production plus Relics.
  ProductionAndRelics
};
```

> Holy shit, have a look at the monk conversion-mechanics!

Don't worry it looks complicated but its mainly only 3 variables, on the wiki is explained more from the code perspective.

Reply:
Looks great now!